mitodl/ocw-hugo-themes
# online / offline detection in offline course sites

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-hugo-themes/issues/847

#### What's this PR do?
In https://github.com/mitodl/ocw-hugo-themes/pull/832, we added a theme for rendering course sites offline for download. This PR expands on that and adds a number of features for detecting if the user is online or offline while viewing the downloaded site. Search links are stripped from text in the course info drawers, some links in the header and footer are hidden entirely, and a message is shown on video pages indicating that an internet connection is required to view the videos.

#### How should this be manually tested?
 - Clone `ocw-hugo-themes` on this PR branch, `ocw-hugo-projects`, and https://github.mit.edu/mitocwcontent/18.01sc-fall-2010
 - Set the following `.env` variables:
```
STATIC_API_BASE_URL=https://ocw.mit.edu/
COURSE_HUGO_CONFIG_PATH=/path/to/ocw-hugo-projects/ocw-course-v2/config-offline.yaml
```
 - Run the following command: `npm run build -- /path/to/mitocwcontent/18.01sc-fall-2010 /path/to/ocw-hugo-projects/ocw-course-v2/config-offline.yaml`
 - Browse to `/path/to/mitocwcontent/18.01sc-fall-2010/dist` and double click on `index.html` to open the course site using some variant of Chrome so we can simulate being offline easily
 - In online mode, verify that:
   - Clicking the "about OCW" etc. links in the header goes to production OCW
   - Clicking any of the search links in the course info drawer on the course home page goes to the search result on production OCW
   - Verify the same thing for the course info drawer in a sub page (syllabus, etc.)
   - Browse to one of the many video pages in 18.01sc and verify that the video plays back and the transcript works
 - Open the developer console, go to the network tab and in the throttling dropdown select "offline" then verify the following
   - On the course home page, links that would normally go somewhere online in the header and footer are hidden
   - The search links in the course info drawers on the course home page as well as sub pages are rendered as plain text
   - On video pages, you should see a message saying "Viewing videos requires an internet connection"
 - Repeat the above with mobile simulation enabled in developer tools

#### Where should the reviewer start?
Check out the newly overridden partials in the `course-offline` theme as well as the links that are hidden in the header and footer. Comment on what you think makes sense to show or hide here when the user is detected as offline. Check out JS functionality that strips links and hides and shows content if you are offline / online in the new partial `hide_offline_links.html`.

#### Any background context you want to provide?
This PR and the ones before it relating to the `course-offline` theme are part of work that will restore downloading a version of an OCW course that is browseable locally on any device. ZIP archives of each course will be generated by a pipeline in Concourse and a future PR will write in links to these ZIP files on the live course sites to make theme available for download.

#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/12089658/191140247-609d3d30-7fc1-464b-9baf-7c2a86b32053.png)
![image](https://user-images.githubusercontent.com/12089658/191140269-2d1e4cbe-b3d6-43ab-a908-556697436ec2.png)
![image](https://user-images.githubusercontent.com/12089658/191140408-c6feff23-42d8-4260-a696-4418977614c1.png)

![image](https://user-images.githubusercontent.com/12089658/191140980-51c93443-36cd-4116-bc16-005daee14242.png)
![image](https://user-images.githubusercontent.com/12089658/191141014-885b4966-15cf-4855-adb6-1bd028f1d80d.png)
![image](https://user-images.githubusercontent.com/12089658/191141043-6bd784c8-90e3-4238-a977-2894a10378d0.png)
